### PR TITLE
client: add shape settings

### PR DESCRIPTION
**Pull request by copycat-killer**
Lets add `set_shape` to `client` as well.

Commits needed for this PR:

- [x] squash commit on size signals;
- [x] add a `property::size` signal in `client.c` to avoid the shape being applied twice;
- [x] add documentation to <s>`wibar.lua` and</s> the API docs.
- [x] add a test.

### CODE REVIEW

**Comment by Elv13** [comment]
Thanks **a lot** for this and *finally :)* enter the official commit log. As discussed on IRC:

 * create a `property::size` for clients in `client.c` to avoid the code being called twice
 * connect the `border_width` and size signals to `client.object._apply_shape`
 * add a documentation block in `wibar.lua`
 * add a test to make CodeCov happy, this can be done in `test-struts.lua` by adding a shape to the wibars there. As they are already programmed to move and resize, the code should be stressed enough

**Comment by psychon** [comment]
Sorry, but this conflicts with `awful.client.shape` which already tries to do *some* shape handling on clients (this code is what makes `xeyes` slightly less ugly). I think a new solution would need to incorporate this handling. How "bad" would it be to have the "actual applying of a shape" in `awful.client.shape` and `awful.client` just calls functions from there? (Or `awful.client.shape` monkey-patches `awful.client` to avoid a `require()`-dependency-loop)

**Comment by Elv13** [comment]
>  How "bad" would it be to have the "actual applying of a shape" in awful.client.shape

Right. `_apply_shape` could become a `local function` and be called in `if c._shape then apply_shape(c); return`. Adding the `property::border_width` will also be required, as found by copycat-killer.

**Comment by copycat-killer** [comment]
Moved the shaping to `awful.client.shape.set`.

@Elv13 I don't understand the relation with `wibar.lua`: why (and where) do I have to add documentation in it?

**Comment by Elv13** [comment]
> @Elv13 I don't understand the relation with wibar.lua: why (and where) do I have to add documentation in it?

Nevermind. I was confused with wibox shapes again. However the problem is real, but forget it. I will create a PR to fix that.

**Comment by copycat-killer** [comment]
So, it seems that documentation is grabbed directly from the code, and I added the proper comments.

Any suggestions about the test?

File: lib/awful/client/shape.lua [hunk 1]
```diff
@@ -51,8 +51,54 @@ function shape.get_transformed(c, shape_name)
     return result
 end
 
+--- Set the client shape.
+-- @function awful.client.shape.set
+-- @client c The client whose shape should be retrieved
+-- @tparam gears.shape A gears.shape compatible function.
+-- @see gears.shape
+function shape.set(c, shape)
```
Comment (Elv13) [inline lib/awful/client/shape.lua]: `function client.object.set_shape` is still required to eixst in `client.lua`. Otherwise `c.shape = ...` wont work

**Comment by Elv13** [comment]
> Any suggestions about the test?

What about `client.get()[1].shape = gears.shape.rounded_rect` in `test-resize.lua:18`. This should apply the shape and resize it in all sorts of ways.

File: lib/awful/client/shape.lua (hunk 1 above)
Comment (copycat-killer) [inline lib/awful/client/shape.lua]: Added.

File: lib/awful/client/shape.lua [hunk 2]
```diff
@@ -87,6 +133,9 @@ capi.client.connect_signal("property::shape_client_bounding", shape.update.bound
 capi.client.connect_signal("property::shape_client_clip", shape.update.clip)
 capi.client.connect_signal("property::width", shape.update.all)
```
Comment (Elv13) [inline lib/awful/client/shape.lua]: If `size` works, then `width` and `height` can be removed, right?

Comment (Elv13) [inline lib/awful/client/shape.lua]: Oh, and it should call `shape.update.all`, not `set_shape`.

Comment (copycat-killer) [inline lib/awful/client/shape.lua]: It seems so.

**Comment by copycat-killer** [comment]
Added test too.

**Comment by Elv13** [comment]
I think you accidentally removed all commits

**Comment by copycat-killer** [comment]
Haha yeah, I moved my mods to a dedicated branch, but forgot that this PR was bound to my master.

Will open a new one.

**Pull request closed without merging**